crates/polars-plan/src/plans/optimizer/simplify_expr/arity.rs

```diff
@@ -0,0 +1,143 @@
+
+            // true AND x => x
+            if matches!(
+                left_ae,
+                AExpr::Literal(lv) if lv.bool() == Some(true)
+            ) && in_filter
+            {
+                // Only in filter as we might change the name from "literal"
+                // to whatever lhs columns is.
+                return Some(right_ae.clone());
+            }
+
+            // x AND true => x
+            if matches!(
+                right_ae,
+                AExpr::Literal(lv) if lv.bool() == Some(true)
+            ) {
+                return Some(left_ae.clone());
+            }
+
+            // x AND false -> false
+            // FIXME: we need an optimizer redesign to allow x & false to be optimized
+            // in general as we can forget the length of a series otherwise.
+            // In filter we allow it as the length is not important there.
+            if (is_scalar_ae(left, expr_arena) | in_filter)
+                && matches!(
+                    right_ae,
+                    AExpr::Literal(lv) if lv.bool() == Some(false)
+                )
+            {
+                return Some(AExpr::Literal(Scalar::from(false).into()));
+            }
+        },
+        O::Or => {
+            let left_ae = expr_arena.get(left);
+            let right_ae = expr_arena.get(right);
+
+            // false OR x => x
+            if matches!(
+                left_ae,
+                AExpr::Literal(lv) if lv.bool() == Some(false)
+            ) && in_filter
+            {
+                // Only in filter as we might change the name from "literal"
+                // to whatever lhs columns is.
+                return Some(right_ae.clone());
+            }
+
+            // x OR false => x
+            if matches!(
+                right_ae,
+                AExpr::Literal(lv) if lv.bool() == Some(false)
+            ) {
+                return Some(left_ae.clone());
+            }
+            // true OR x => true
+            // FIXME: we need an optimizer redesign to allow true | x to be optimized
+            // in general as we can forget the length of a series otherwise.
+            // In filter we allow it as the length is not important there.
+            if (is_scalar_ae(left, expr_arena) | in_filter)
+                && matches!(
+                    right_ae,
+                    AExpr::Literal(lv) if lv.bool() == Some(true)
+                )
+            {
+                return Some(AExpr::Literal(Scalar::from(true).into()));
+            }
+
+            // x OR true => true
+            // FIXME: we need an optimizer redesign to allow true | x to be optimized
+            // in general as we can forget the length of a series otherwise.
+            // In filter we allow it as the length is not important there.
+            if matches!(
+                left_ae,
+                    AExpr::Literal(lv) if lv.bool() == Some(true)
+            ) && (is_scalar_ae(right, expr_arena) | in_filter)
```

Comment:
Not only apply for literals, but for all scalars. If we are in a filter we can always do this.